src/spatial/transform/CounterBitwidthTransformer.scala

```diff
@@ -11,29 +12,70 @@ import spatial.metadata.control._
 import spatial.metadata.memory._
 import spatial.metadata.blackbox._
 
+import utils.math.log2Up
+
+import emul.FixedPoint
+
 case class CounterBitwidthTransformer(IR: State) extends MutateTransformer
   with AccelTraversal {
 
-  private def optimizeCtrs(ctrs: Seq[Sym[_]]): Seq[Sym[_]] = {
+  /** Calculate the least bitwidth required for an integer. */
+  private def getBitwidth(x: Int): Int = log2Up(x.abs)
+
+  /** Extract the content from Const and cast it to Int. */
+  private def constToInt(x: Sym[_]): Int = 
+    if (x.isConst)
+      x.c.get.asInstanceOf[FixedPoint].toInt
+    else
+      throw new Exception(s"$x is not a Const.")
+
+  /** Create a new CounterNew object with compact bitwidth. */
+  private def getOptimizedCounterNew(ctr: CounterNew[_]): CounterNew[_] = ctr match {
+    case CounterNew(start, stop, step, par) =>
+      // we take the largest magnitude of start and stop to decide the boundary of bitwidth
+      val begin = constToInt(start)
+      val end = constToInt(stop)
+      val bitwidth = math.max(getBitwidth(begin), getBitwidth(end))
+
+      // TODO: Find a better way that can map bitwidth to the exact Fix type
+      if (bitwidth <= 7) {
```

Comment:
@mattfel1 Hi there, I feel that it would be tedious to implement all the bitwidth-to-type cast and there should be a better way that I'm not aware of. Maybe you've met this scenario before and have a good way to deal with it? Thanks!

Reply:
Unfortunately, I don't think there is a non-tedious way to do this. 
 Since each bitwidth is its own trait (argon/lang/types/CustomBitWidths.scala), its painful to work with.  Some people have used quasiquotes for this problem before but there isn't a nice way that I know of.

Reply:
Thanks @mattfel1 ! In my latest update I manually added all the mappings for different bit-width values. Hope it looks fine.